src/vision/status_quo/barbara_carefully_dismisses_embedded_future.md

```diff
@@ -0,0 +1,210 @@
+could function as a lightweight form of multithreading. To support this, Barbara
+decides to implement an executor designed to function like a background thread.
+This executor is started using a `spawn` function, and otherwise runs entirely
+in kernel callbacks.
+
+Barbara identifies several factors that add branching and error handling code to
+the executor:
+
+1. `spawn` should be a safe function, because it is called by high-level
+   application code. However, that means it can be called by the future it
+   contains. If handled naively, this would result in dropping the future while
+   it executes. Barbara adds runtime checks to identify this situation.
+1. `Waker` is `Sync`, so on a multithreaded system, a future could give another
+   thread access to its `Waker` and the other thread could wake it up. This
+   could happen while the `poll` is executing, before `poll` returns
+   `Poll::Pending`. Therefore, Barbara concludes that if `wake` is called while
+   a future is being polled then the future should be re-polled, even if the
+   current `poll` returns `Poll::Pending`. This requires putting a retry loop
+   into the executor.
+1. A kernel callback may call `Waker::wake` after its future finishes executing,
+   and the executor must ignore those wakeups. This duplicates the "ignore
+   spurious wakeups" functionality that exists in the future itself.
+
+Ultimately, this made the executor logic [quite
+nontrivial](https://github.com/tock/design-explorations/blob/master/size_comparison/futures/src/task.rs).
+The executor logic can be monomorphized for each future, which allows the
+compiler to make inlining optimizations, but results in a significant amount of
+duplicate code. Alternatively, it could be adapted to use function pointers or
+vtables to avoid the code duplication, but then the compiler *definitely* cannot
+inline `Future::poll` into the kernel callbacks.
+
+At this point, Barbara realizes she needs to do some benchmarking. Barbara comes
+up with a sample application -- an app that blinks a led and responds to button
+presses -- and implements it twice. One implementation does not use `Future` at
+all, the other does. In addition to two async drivers (a timer driver and a GPIO
+driver), the `Future`-based app uses Barbara's executor as well as a single
+combinator (that manages the state of the app).
+
+Barbara publishes an
+[analysis](https://github.com/tock/design-explorations/tree/master/size_comparison)
+of the relative sizes of the two app implementations, finding a large percentage
+increase in both code size and RAM usage (note: stack usage was not
+investigated). Barbara concludes that **`core::future::Future` is not suitable
+for highly-resource-constrained environments** due to the amount of code and RAM
+required to implement executors and combinators.
+
+Barbara redesigns the library she is building to use a [different
+concept](https://github.com/tock/design-explorations/tree/master/zst_pointer_async)
```

Comment:
Can you talk more about what design decisions this implementation made that made it more light-weight?

Reply:
I added 2 things to address this:

1. An analysis of why the `Future`-based app is larger than the no-`Future` app (currently lines 174-200)
2. A list below (currently lines 214-220) explaining how I (Barbara) avoided the same costs in my async traits.